snap/info_snap_yaml.go

```diff
@@ -494,46 +535,70 @@ func setHooksFromSnapYaml(y snapYaml, snap *Info, strk *scopedTracker) {
 		}
 
 		snap.Hooks[hookName] = hook
+
 		// Bind all plugs/slots listed in this hook
-		for _, plugName := range yHook.PlugNames {
-			plug, ok := snap.Plugs[plugName]
-			if !ok {
-				// Create implicit plug definitions if required
-				plug = &PlugInfo{
-					Snap:      snap,
-					Name:      plugName,
-					Interface: plugName,
-					Hooks:     make(map[string]*HookInfo),
-				}
-				snap.Plugs[plugName] = plug
+		bindPlugsToHook(hook, yHook.PlugNames, snap, strk)
+		bindSlotsToHook(hook, yHook.SlotNames, snap, strk)
+	}
+}
+
+func bindSlotsToHook(hook *HookInfo, slotNames []string, snap *Info, strk *scopedTracker) {
+	for _, slotName := range slotNames {
+		slot, ok := snap.Slots[slotName]
+		if !ok {
+			// Create implicit slot definitions if required
+			slot = &SlotInfo{
+				Snap:      snap,
+				Name:      slotName,
+				Interface: slotName,
+				Hooks:     make(map[string]*HookInfo),
 			}
-			// Mark the plug as scoped.
-			strk.markPlug(plug)
-			if plug.Hooks == nil {
-				plug.Hooks = make(map[string]*HookInfo)
+			snap.Slots[slotName] = slot
+		}
+
+		// Mark the slot as scoped.
+		strk.markSlot(slot)
+
+		if slot.Hooks == nil {
+			slot.Hooks = make(map[string]*HookInfo)
+		}
+		hook.Slots[slotName] = slot
+		slot.Hooks[hook.Name] = hook
+	}
+}
+
+func bindPlugsToHook(hook *HookInfo, plugNames []string, snap *Info, strk *scopedTracker) {
+	for _, plugName := range plugNames {
+		plug, ok := snap.Plugs[plugName]
+		if !ok {
+			// Create implicit plug definitions if required
+			plug = &PlugInfo{
+				Snap:      snap,
+				Name:      plugName,
+				Interface: plugName,
+				Hooks:     make(map[string]*HookInfo),
 			}
-			hook.Plugs[plugName] = plug
-			plug.Hooks[hookName] = hook
+			snap.Plugs[plugName] = plug
 		}
-		for _, slotName := range yHook.SlotNames {
-			slot, ok := snap.Slots[slotName]
-			if !ok {
-				// Create implicit slot definitions if required
-				slot = &SlotInfo{
-					Snap:      snap,
-					Name:      slotName,
-					Interface: slotName,
-					Hooks:     make(map[string]*HookInfo),
-				}
-				snap.Slots[slotName] = slot
+
+		// Mark the plug as scoped.
+		strk.markPlug(plug)
+
+		hook.Plugs[plug.Name] = plug
+
+		// if the hook has a component, then we know that this hook should be
+		// listed as a component hook in the plug. otherwise, it is a normal
+		// snap hook.
+		if hook.Component != nil {
+			if plug.ComponentHooks == nil {
+				plug.ComponentHooks = make(map[string]*HookInfo)
 			}
-			// Mark the slot as scoped.
-			strk.markSlot(slot)
-			if slot.Hooks == nil {
-				slot.Hooks = make(map[string]*HookInfo)
+			plug.ComponentHooks[hook.Name] = hook
```

Comment:
as I said I don't understand this mapping as we lose from which component the hook is for